woocommerce/woocommerce-android
# Feat/products 2707 simple product details add images

## RESOURCES
[The Figma file](https://www.figma.com/file/QAkdm5HZdLlG0b8vUkv1pX/Products---Add-%26-Edit-Products?node-id=460%3A1760&viewport=-1704%2C1577%2C0.5334523320198059) and [the initial issue](https://github.com/woocommerce/woocommerce-android/issues/2707)

## CHANGES / DESCRIPTION

**EDIT**:
This PR will contain the following:
- added the **Add image** flow for the product **"add"** path

As pointed out by @0nko, the split between the product **edit** and **add** paths was getting too convoluted and complicated to manage/follow. A follow up [PR here](https://github.com/GuilhermeMota93/woocommerce-android/pull/2) (in a waiting status) was going to introduce some refactor work in order to help manage better the 2 states of a `Product` flow.

Given the initial review in this PR, it made sense to bring that work sooner and even refactor things a bit more so that the `productDraft` is the only source of truth. This helped in cleaning up quite a bit and keeping things in line with other existing work 💪 

The biggest changes introduced are:

**1)** the `start()` function of the `ProductDetailViewModel` that, for the **add** path, now loads the view state with (as suggested!) a **default product.**
```
private fun startAddNewProduct() {
  val preferredSavedType = AppPrefs.getSelectedProductType()
  val defaultProductType = ProductType.fromString(preferredSavedType)
  viewState = viewState.copy(productDraft = ProductHelper.getDefaultNewProduct(type = defaultProductType))
}
```
This **default product** takes in account the current preferred product type of the user, from the `AppPrefs`, and sets all of the needed information to its default values inside `ProductHelper` - **it felt like the best place to put it, but let me know if you see it best somewhere else.**

**2)** the slight modification in behaviour of the `onEventMainThread` of the `OnProductImagesUpdateCompletedEvent`; for the **add** path, bypasses the fetching of a remote product (as this new product doesn't exist yet) and straight away triggers the `checkImageUploads` to keep the UI state of the image displaying properly 👌 

This suggested approach also brought in, as per the already existing view state observe states, the different steps for a product:
- title
- description
- the cards

These will need some tweaking/cleanup that will follow in upcoming PRs

As always, let me know your thoughts on it 🙏 

## TESTING
- Click on the PRODUCT bottom nav bar button
    - Wait for the screen to finish loading
    - Click on the FAB button
    - Notice the product detail for a new product flow, empty with some default views
    - Click top image container
    - Notice the Image Picker screen showing
    - Select `Add Photos`
    - Notice the select input alert/dialog
 - Scenario 1:
    - Select `Choose from device`
    - Notice the camera/device roll displaying the available images
    - Select one (or more) image(s)
 - Scenario 2:
    - Select `Take photo`
    - Notice the camera opening to take a photo
    - Take and select photo
 - Scenario 3:
    - Select `WordPress media library`
    - Notice the WordPress library displaying all available images
    - Select one (or more) image(s)
- Notice the image showing in the list
- Press back to navigate to the Product Detail screen 
- Notice the horizontal list with the selected images displaying

## SCREENSHOTS

<image src="https://user-images.githubusercontent.com/7058393/91103120-a05f3900-e662-11ea-9a87-988794b5fb4f.gif" width="320"/><image src="https://user-images.githubusercontent.com/7058393/91103125-a7864700-e662-11ea-93fe-e30471a69a0d.gif" width="320"/><image src="https://user-images.githubusercontent.com/7058393/91103131-ab19ce00-e662-11ea-90bd-4645ba70b9df.gif" width="320"/>
